src/Dialect/ONNX/Transforms/QDQAroundOpOpt.cpp

```diff
@@ -0,0 +1,132 @@
+#include <src/Dialect/ONNX/ONNXOps.hpp>
+#include <src/Dialect/ONNX/ONNXOps/OpHelper.hpp>
+
+using namespace mlir;
+using namespace onnx_mlir;
+std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+    ONNXTransposeOp transposeOp) {
+  return {transposeOp.getData(), transposeOp.getTransposed()};
+}
+std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+    ONNXUnsqueezeOp unsqueezeOp) {
+  return {unsqueezeOp.getData(), unsqueezeOp.getExpanded()};
+}
+std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+    ONNXSqueezeOp squeezeOp) {
+  return {squeezeOp.getData(), squeezeOp.getSqueezed()};
+}
+std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+    ONNXReshapeOp reshapeOp) {
+  return {reshapeOp.getData(), reshapeOp.getReshaped()};
+}
+std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+    ONNXGatherOp gatherOp) {
+  return {gatherOp.getData(), gatherOp.getOutput()};
+}
+// std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+//     ONNXReduceSumOp reduceOp) {
+//   return {reduceOp.getData(), reduceOp.getReduced()};
+// }
+std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+    ONNXSliceOp sliceOp) {
+  return {sliceOp.getData(), sliceOp.getOutput()};
+}
+std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+    ONNXResizeOp resizeOp) {
+  return {resizeOp.getX(), resizeOp.getY()};
+}
+std::tuple<Value /*input*/, Value /*output*/> getDataInputOutput(
+    ONNXFlattenOp flattenOp) {
+  return {flattenOp.getInput(), flattenOp.getOutput()};
+}
+namespace {
+template <typename T>
+class RemoveQDQAroundOpPattern : public OpRewritePattern<T> {
+public:
+  using OpRewritePattern<T>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      T op, PatternRewriter &rewriter) const override {
+    Value input, output;
+    std::tie(input, output) = getDataInputOutput(op);
+
+    auto dqOp = input.getDefiningOp<ONNXDequantizeLinearOp>();
+    if (!dqOp) {
+      return failure();
+    }
+    if (output.hasOneUse()) {
+      Operation *firstOp = *(output.getUsers().begin());
+      if (mlir::isa<ONNXQuantizeLinearOp>(firstOp)) {
+        auto qOp = ::llvm::dyn_cast<ONNXQuantizeLinearOp>(firstOp);
+        if (!qOp)
+          return failure();
+        if (!isDequantQuantSame(dqOp, qOp))
+          return failure();
+
+        // Map dqOp inputs to dqOp's inputs
+        IRMapping irMapping;
+        irMapping.map(dqOp, dqOp.getX());
+
+        SmallVector<Value> newInputs;
+        transform(op->getOperands(), std::back_inserter(newInputs),
+            [&](Value operand) { return irMapping.lookupOrDefault(operand); });
+
+        rewriter.replaceOp(dqOp, dqOp.getX());
+        auto newOp = rewriter.create<T>(op.getLoc(),
+            TypeRange{qOp.getResult().getType()}, ValueRange{newInputs},
+            op->getAttrs());
+        rewriter.replaceOp(op, newOp);
+        rewriter.replaceOp(qOp, newOp.getResult());
+        return success();
```

Comment:
I think you can simply the rewrite here a bit:
```suggestion
        auto newOp = rewriter.create<T>(op.getLoc(),
            TypeRange{qOp.getResult().getType()}, ValueRange{dqOp.getX()},
            op->getAttrs());
        rewriter.replaceOp(qOp, newOp.getResult());
        // rewriter.eraseOp(op). // This is optional, DCE will remove it anyway
        return success();
```
This will also handle the case correctly, where the qOp has more than one use

Reply:
some ops have more than 1 input, I need the new X input from dq and other unchanged inputs. Yes, it's not needed to erase/replace orginal op.